corehq/apps/userreports/pillow.py

```diff
@@ -208,6 +211,117 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+            try:
+                table.handle_exception(doc, e)
+            except UserReportsWarning:
+                # remove it until the next bootstrap call
+                self.table_adapters_by_domain[domain].remove(table)
+
+    def process_changes_chunk(self, pillow_instance, changes):
+        """
+        Update UCR tables in bulk by breaking up changes per domain per UCR table.
+            If an exception is raised in bulk operations of a set of changes,
+            those changes are returned to pillow for serial reprocessing.
+        """
+        self.bootstrap_if_needed()
+        # break up changes by domain
+        changes_by_domain = defaultdict(list)
+        for change in changes:
+            # skip if no domain or no UCR tables in the domain
+            if change.metadata.domain and change.metadata.domain in self.table_adapters_by_domain:
+                changes_by_domain[change.metadata.domain].append(change)
+
+        retry_changes = set()
+        for domain, changes_chunk in six.iteritems(changes_by_domain):
+            failed = self._process_chunk_for_domain(domain, changes_chunk)
+            retry_changes.update(failed)
+
+        return retry_changes
+
+    def _get_docs(self, changes):
+        # get docs in bulk and validate revision
+        # break up by doctype
+        _changes_by_doc_store = defaultdict(list)
+        for change in changes:
+            _changes_by_doc_store[change.document_store].append(change)
+
+        # query
+        docs = []
+        for doc_store, _changes in six.iteritems(_changes_by_doc_store):
+            docs.extend(list(doc_store.iter_documents([change.id for change in _changes])))
+
+        # catch missing docs
+        retry_changes = set()
+        docs_by_id = {doc['_id']: doc for doc in docs}
+        for change in changes:
+            if change.id not in docs_by_id:
+                # we need to capture DocumentMissingError which is not possible in bulk
+                #   so let pillow fall back to serial mode to capture the error for missing docs
+                retry_changes.add(change)
+                continue
+            try:
+                ensure_matched_revisions(change, docs_by_id.get(change.id))
+            except DocumentMismatchError:
+                retry_changes.add(change)
+        good_changes = set(changes) - retry_changes
+        return retry_changes, [docs_by_id.get(change.id) for change in good_changes]
+
+    def _process_chunk_for_domain(self, domain, changes_chunk):
+        adapters = self.table_adapters_by_domain[domain]
+
+        to_delete_by_adapter = defaultdict(list)
+        rows_to_save_by_adapter = defaultdict(list)
+        async_configs_by_doc_id = defaultdict(list)
+        to_update = {change for change in changes_chunk if not change.deleted}
+        retry_changes, docs = self._get_docs(to_update)
+
+        for doc in docs:
+            eval_context = EvaluationContext(doc)
+            for adapter in adapters:
```

Comment:
You'll need to make a copy of this list in order to avoid the issue I described above about skipping items in the list